### PR TITLE
Correct offsets in dual arm husky URDF

### PR DESCRIPTION
**Pull request by wxmerkt**
We've noticed a few issues with the dual arm husky URDF:

1. The mass for the batteries and computers isnt included in the base mass (not part of this PR - this affects the Gazebo simulation, cf. https://github.com/husky/husky_simulator/issues/27#issuecomment-258621896)
2. The offset from the force torque sensor to the gripper palms seems to be off - this is why the hands are "floating" in gazebo and rviz and other visualisers. It is set to 8cm off the hand-side of the force torque sensor. The data sheet doesn't give the palm offset, but we've extracted it from the palm CAD model.
3. The offset for the PTU mount was wrong - we've remeasured with measuring tape (20cm) and extracted the correct offset from the CAD model of the bulkhead (19.876cm). This change stops the PTU from floating. 
3. The checked in/committed URDF doesn't reflect the most recent xacro file, i.e. the PTU axes were off. We've removed the checked-in URDF to force users to generate from xacro and added a convenience script. We've also removed an autosave artefact.

### CODE REVIEW

**Comment by wxmerkt** [comment]
@TheDash Do these modifications look OK and are okay to merge into the dual_ur5_husky branch? We've also noticed the grippers aren't mounted the way they are described in the URDF, there is some offset, maybe 45 degrees?

File: husky_control/CMakeLists.txt [hunk 1]
```diff
@@ -7,8 +7,8 @@ catkin_package()
 
 roslaunch_add_file_check(launch)
 
-catkin_install_python(PROGRAMS scripts/stow_ur5 scripts/stow_ur5_left scripts/stow_ur5_right DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})
+install(PROGRAMS scripts/stow_ur5 scripts/stow_ur5_left scripts/stow_ur5_right DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION})
```
Comment (UltronDestroyer) [inline husky_control/CMakeLists.txt:10]: What's the reason for this change?

Putting it in bin allows it to be run with rosrun, what happens now?

File: husky_description/urdf/description.xacro [hunk 2]
```diff
@@ -31,7 +31,7 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
   <xacro:arg name="kinect_enabled" default="false" />
   <xacro:arg name="robotiq_grippers_enabled" default="$(optenv ROBOTIQ_GRIPPERS_ENABLED false)" />
  
-  <xacro:include filename="$(find husky_description)/urdf/husky.urdf.xacro" />
+  <xacro:include filename="$(find husky_description)/urdf/dual_arm_husky.urdf.xacro" />
```
Comment (UltronDestroyer) [inline husky_description/urdf/description.xacro:34]: I've written the code inside the husky.urdf.xacro such that it is parametrized, with hopes of one day merging it into the main branch. Launch files will look for husky.urdf.xacro and so I do not want to make shotgun code.

File: husky_description/urdf/dual_arm_husky.urdf.xacro [hunk 3]
```diff
@@ -98,9 +98,9 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
   <!-- Interial link stores the robot's inertial information -->
   <link name="inertial_link">
     <inertial>
-      <mass value="33.455"/>
-      <origin xyz="-0.08748 -0.00085 0.09947"/>
-      <inertia ixx="0.6022" ixy="-0.02364" ixz="-0.1197" iyy="1.7386" iyz="-0.001544" izz="2.0296"/>
+      <mass value="48.39054"/>
+      <origin xyz="-0.0648 -0.00059 0.07895"/>
+      <inertia ixx="2.1634275613800003" ixy="-0.055452352110000006" ixz="-0.41083822026" iyy="3.5314689495400002" iyz="-0.004366145530000001" izz="4.4759585875" />
```
Comment (UltronDestroyer) [inline husky_description/urdf/dual_arm_husky.urdf.xacro:103]: Out of completeness sake - how were these values determined?

File: husky_description/urdf/generate_dual_arm_husky_urdf.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+rosrun xacro xacro.py dual_arm_husky.urdf.xacro laser_enabled:=false ur5_enabled:=false kinect_enabled:=false dual_ur5_enabled:=true robotiq_grippers_enabled:=true > dual_arm_husky.urdf
```
Comment (UltronDestroyer) [inline husky_description/urdf/generate_dual_arm_husky_urdf.sh:2]: I like this. Shows users how to build it. Not how we usually do things - but I like it.

**Comment by UltronDestroyer** [comment]
Thanks @wxmerkt 

This PR is excellent. I've commented on some things regarding semantics and small details - but the whole bulk of it is okay. 

Will merge once they're touched up!

File: husky_control/CMakeLists.txt (hunk 1 above)
Comment (wxmerkt) [reply husky_control/CMakeLists.txt:10]: I've added this today after original PR - it's labelled as an odd change; it's for some reason necessary for catkin tools to work correctly - otherwise it doesn't copy the files into the correct install workspace

File: husky_description/urdf/description.xacro (hunk 2 above)
Comment (wxmerkt) [reply husky_description/urdf/description.xacro:34]: I agree - but as long as it's not being merged up into the general master branch and the husky fixed up (i.e. the elements from dual_arm_husky merged in), this is the only way to make Gazebo happy

File: husky_description/urdf/dual_arm_husky.urdf.xacro (hunk 3 above)
Comment (wxmerkt) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: We've contacted Clearpath Support and Martin Cote sent us these values. The axes/frame were wrong and I've changed them by inspecting the URDF. It's much better, but the Husky still doesn't drive in Gazebo and is highly unstable. Maybe they are off? I'd be much obliged if you could take a look at it/from the CAD files and figure out whether other values should be put in here

**Comment by wxmerkt** [comment]
Thanks for the review @TheDash. The CMakeLists.txt change is an odd one, but required for installation with catkin tools to work (it goes against any docs as well but files otherwise don't get copied to the correct location). It should be reverted once other behaviour with catkin tools has been shown to work. While the URDF changes here fix a lot for us, the remaining issues are:
1. The orientation of the mounting of the grippers to the FTS is off, maybe 45 degrees?
2. The Gazebo simulation is still not working, the Husky performs a wheelie and keeps rocking on its backwheels. Driving forwards looks OK in rviz but not in the gazebo client, driving backwards doesnt work at all

File: husky_description/urdf/description.xacro (hunk 2 above)
Comment (wxmerkt) [reply husky_description/urdf/description.xacro:34]: Hi @TheDash - this omission/not-integration just killed our Husky, cf. support request #4137 (we spoke to Martin). I've manually made this change to the provided husky installation, rebuilt and it solved our problem (it lost comms due to a malformed URDF - base_laser was not unique causing the husky_node to crash)

File: husky_description/urdf/dual_arm_husky.urdf.xacro (hunk 3 above)
Comment (UltronDestroyer) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: My understanding is this has been fixed and we determined the mass was without the battery accounted for? Could you confirm?

Edit: just read above comments. Ignore this reply.

Comment (wxmerkt) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: Yes, my best guess would be that batteries and computers were not accounted for. I obtained the values as seen in this diff from Martin, which made Gazebo work slightly better but it's still not navigating/driving. Were you able to run the Gazebo simulation with the dual arm husky and use navigation/move_base with goals given via rviz?

File: husky_description/urdf/dual_arm_husky.urdf.xacro [hunk 5]
```diff
@@ -3036,7 +3036,7 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
   <!-- end of joint list -->
   <!-- joint attaching parent link to robotiq hand -->
   <joint name="r_robotiq_hand_joint" type="fixed">
-    <origin rpy="1.57 0 0" xyz="0 0 0.05398"/>
+    <origin rpy="1.57 0 2.355" xyz="0 0 0.05398"/>
```
Comment (wxmerkt) [inline husky_description/urdf/dual_arm_husky.urdf.xacro:3039]: @TheDash We've approximately guessed these rotation values from how the grippers were oriented in visualisation and in reality. Are there any documents on how they should have been mounted during installation? Is such a variation in rotation possible or are there physical alignment points? Both hands are "off" to varying degrees

Comment (UltronDestroyer) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:3039]: @jeff-o 

They should've been mounted in the visualized model to be an exact copy of how it was mounted in reality.  If they are not correct - it is highly likely that I missed this bug before it was shipped out.

File: husky_description/urdf/description.xacro (hunk 2 above)
Comment (UltronDestroyer) [reply husky_description/urdf/description.xacro:34]: Sorry @wxmerkt for the delays, my email has been marking these as read for some reason.

The way the husky is built is that everything goes through the  husky.urdf.xacro file. husky_gazebo calls the xacro.py that rebuilds the urdf every time from the husky.urdf.xacro, is there any reason it had to be renamed?

I'm not entirely sure what's the issue here. Is there a dual_arm_husky.urdf.xacro on the robot that is not on this repo somehow?

File: husky_control/CMakeLists.txt (hunk 1 above)
Comment (UltronDestroyer) [reply husky_control/CMakeLists.txt:10]: Does the rosrun command portion of it still work? Can it still find the binaries while in the share destination?

File: husky_description/urdf/dual_arm_husky.urdf.xacro (hunk 3 above)
Comment (UltronDestroyer) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: I never did attempt a full navigation in Rviz to be honest. I tested the arms to see if they could navigate and move, as well as the grippers. We do not have a continous integration setup that allows us to test if all of our features are still working after something new as added, so it seems that what we are seeing here is the result of that ( move_base/navigation addition ).

File: husky_description/urdf/description.xacro (hunk 2 above)
Comment (wxmerkt) [reply husky_description/urdf/description.xacro:34]: Hi @TheDash - the issue is that there is both a husky.urdf.xacro and a dual_arm_husky.urdf.xacro in this repository/branch and on the robot. The one that appears to be the most recent is the dual_arm_husky.urdf.xacro - when compiled with xacro it works. The one compiled from husky.urdf.xacro crashes. I believe a solution for this would be to either accept this change or to alternatively integrate dual_arm_husky.urdf.xacro into husky.urdf.xacro (cleaner).

File: husky_description/urdf/dual_arm_husky.urdf.xacro (hunk 3 above)
Comment (wxmerkt) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: On that note, move_base (ros-indigo-navigation) doesn't appear to be installed on the physical robot - did you use navigation on the husky before shipping? The grippers also don't work in gazebo for us, but we are seeing some robotiq gazebo plugins on the husky itself - which repositories are these from?

File: husky_description/urdf/description.xacro (hunk 2 above)
Comment (UltronDestroyer) [reply husky_description/urdf/description.xacro:34]: Hi @wxmerkt,

My apologies - I was looking at the wrong branch. It's been many months since I've done this ;p 

It does appear that this change is correct and we can go forward with it.

File: husky_description/urdf/dual_arm_husky.urdf.xacro (hunk 3 above)
Comment (UltronDestroyer) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:103]: That is correct - husky_navigation was not used on the robot before shipping. The robotiq grippers should work in gazebo. They do not move with ros_control however, they move using direct joint state publisher commands. 

Are you seeing the plugins via source or debs? If its debs its most likely the ones distributed from robotiq themselves - if its by source you can check doing "git remote show origin"

File: husky_control/CMakeLists.txt (hunk 1 above)
Comment (UltronDestroyer) [reply husky_control/CMakeLists.txt:10]: I believe this is the only thing holding up the PR @wxmerkt

Comment (wxmerkt) [reply husky_control/CMakeLists.txt:10]: Yes, it did work with Gazebo. Haven't tested it on the real robot as I haven't pushed this bunch of changes onto the robot yet (bar the description.xacro one), and we aren't using catkin-tools on the robot yet.

**Comment by UltronDestroyer** [comment]
LGTM.

File: husky_description/urdf/dual_arm_husky.urdf.xacro [hunk 6]
```diff
@@ -482,7 +482,7 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
   </joint>
   <!-- Connect PTU to Husky bulkhead -->
   <joint name="husky_ptu_joint" type="fixed">
-    <origin rpy="0 0 0" xyz="0.365 0 0.221"/>
```
Comment (wxmerkt) [inline husky_description/urdf/dual_arm_husky.urdf.xacro:485]: @TheDash how did you originally come up with .221? We have just noticed the bracket that mounts the Bumblebee onto the PTU isn't incorporated in the model at all and hence the centre of rotation is off by ~3cm causing all images not to align :(

Comment (wxmerkt) [reply husky_description/urdf/dual_arm_husky.urdf.xacro:485]: Ah, I figured it out! This offset which caused the whole PTU to float needed to be added to the mount_link for the bumblebee to be placed correctly. The value I put in is correct for this one, I will add another PR with the corrected mount link once I've figured the other two offsets out
